pytorch_lightning/trainer/trainer.py

```diff
@@ -852,7 +892,72 @@ def _pre_training_routine(self):
         self.on_pretrain_routine_end()
         ref_model.on_pretrain_routine_end()
 
+    def reset_train_val_dataloaders(self, model) -> None:
+        """
+        Resets train and val dataloaders if none are attached to the trainer.
+
+        The val dataloader must be initialized before training loop starts, as the training loop
+        inspects the val dataloader to determine whether to run the evaluation loop.
+        """
+        if self.train_dataloader is None:
+            self.reset_train_dataloader(model)
+
+        if self.val_dataloaders is None:
+            self.reset_val_dataloader(model)
```

Comment:
before the loop refactor, i think we should refactor the dataloading dependencies here first: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/data_loading.py
 
it seems like we should be doing:
- ensure datamodule.on_load_checkpoint has finished running as this could update the dataloader state
- then initialize the dataloaders
- then extract out the dataloader lengths
- then filter down to the number of batches needed to run and when to check validation

these last 3 seem like things that should happen in the start of the training loop, and we can share utility funtions across loops for getting dataloader lengths and filtering based on `limit_*_batches` arguments to the trainer. what do you think?

Reply:
the current loop is designed to work exactly as the old one. So from my perspective it would be fine to keep it for now and keep this in a backlog of this loop where we want to revisit/further refactor each loop individually after pushing everything into one PR.

If we change behaviour too much now, chances are that we introduce some bug we don't know about because we discard the tests due to the behavioral changes

Reply:
agree with Justus

Absolutely, the dataloader handling can be standardized across loops as can many other common helper utility functions. However, the refactor approach we currently take is top down, i.e., move things in the right place first and then iterate on a finer level.